File: arch/arch.cc
#include "lib/core/globals.h"
#include "lib/encoders/encoders.h"
#include "lib/decoders/decoders.h"
#include "lib/config/config.h"
#include "arch/agat/agat.h"
#include "arch/aeslanier/aeslanier.h"
#include "arch/amiga/amiga.h"
#include "arch/apple2/apple2.h"
#include "arch/brother/brother.h"
#include "arch/c64/c64.h"
#include "arch/f85/f85.h"
#include "arch/fb100/fb100.h"
#include "arch/ibm/ibm.h"
#include "arch/macintosh/macintosh.h"
#include "arch/micropolis/micropolis.h"
#include "arch/mx/mx.h"
#include "arch/northstar/northstar.h"
#include "arch/rolandd20/rolandd20.h"
#include "arch/smaky6/smaky6.h"
#include "arch/tartu/tartu.h"
#include "arch/tids990/tids990.h"
#include "arch/victor9k/victor9k.h"
#include "arch/zilogmcz/zilogmcz.h"
#include "arch/arch.h"

std::unique_ptr<Encoder> Arch::createEncoder(Config& config)
{
    if (!config.hasEncoder())
        error("no encoder configured");
    return createEncoder(config->encoder());
}

std::unique_ptr<Encoder> Arch::createEncoder(const EncoderProto& config)
{
    static const std::map<int,
        std::function<std::unique_ptr<Encoder>(const EncoderProto&)>>
        encoders = {
            {EncoderProto::kAgat,       createAgatEncoder       },
            {EncoderProto::kAmiga,      createAmigaEncoder      },
            {EncoderProto::kApple2,     createApple2Encoder     },
            {EncoderProto::kBrother,    createBrotherEncoder    },
            {EncoderProto::kC64,        createCommodore64Encoder},
            {EncoderProto::kIbm,        createIbmEncoder        },
            {EncoderProto::kMacintosh,  createMacintoshEncoder  },
            {EncoderProto::kMicropolis, createMicropolisEncoder },
            {EncoderProto::kNorthstar,  createNorthstarEncoder  },
            {EncoderProto::kTartu,      createTartuEncoder      },
            {EncoderProto::kTids990,    createTids990Encoder    },
            {EncoderProto::kVictor9K,   createVictor9kEncoder   },
    };

    auto encoder = encoders.find(config.format_case());
    if (encoder == encoders.end())
        error("no encoder specified");

    return (encoder->second)(config);
}

std::unique_ptr<Decoder> Arch::createDecoder(Config& config)
{
    if (!config.hasDecoder())
        error("no decoder configured");
    return createDecoder(config->decoder());
}

std::unique_ptr<Decoder> Arch::createDecoder(const DecoderProto& config)
{
    static const std::map<int,
        std::function<std::unique_ptr<Decoder>(const DecoderProto&)>>
        decoders = {
            {DecoderProto::kAgat,       createAgatDecoder       },
            {DecoderProto::kAeslanier,  createAesLanierDecoder  },
            {DecoderProto::kAmiga,      createAmigaDecoder      },
            {DecoderProto::kApple2,     createApple2Decoder     },
            {DecoderProto::kBrother,    createBrotherDecoder    },
            {DecoderProto::kC64,        createCommodore64Decoder},
            {DecoderProto::kF85,        createDurangoF85Decoder },
            {DecoderProto::kFb100,      createFb100Decoder      },
            {DecoderProto::kIbm,        createIbmDecoder        },
            {DecoderProto::kMacintosh,  createMacintoshDecoder  },
            {DecoderProto::kMicropolis, createMicropolisDecoder },
            {DecoderProto::kMx,         createMxDecoder         },
            {DecoderProto::kNorthstar,  createNorthstarDecoder  },
            {DecoderProto::kRolandd20,  createRolandD20Decoder  },
            {DecoderProto::kSmaky6,     createSmaky6Decoder     },
            {DecoderProto::kTartu,      createTartuDecoder      },
            {DecoderProto::kTids990,    createTids990Decoder    },
            {DecoderProto::kVictor9K,   createVictor9kDecoder   },
            {DecoderProto::kZilogmcz,   createZilogMczDecoder   },
    };

    auto decoder = decoders.find(config.format_case());
    if (decoder == decoders.end())
        error("no decoder specified");

    return (decoder->second)(config);
}

File: java/com/cowlark/fluxengine/arch/Arch.java
package com.cowlark.fluxengine.arch;

import com.cowlark.fluxengine.arch.agat.AgatDecoder;
import com.cowlark.fluxengine.arch.aeslanier.AesLanierDecoder;
import com.cowlark.fluxengine.arch.amiga.AmigaDecoder;
import com.cowlark.fluxengine.arch.apple2.Apple2Decoder;
import com.cowlark.fluxengine.arch.brother.BrotherDecoder;
import com.cowlark.fluxengine.arch.c64.Commodore64Decoder;
import com.cowlark.fluxengine.arch.f85.DurangoF85Decoder;
import com.cowlark.fluxengine.arch.fb100.Fb100Decoder;
import com.cowlark.fluxengine.arch.ibm.IbmDecoder;
import com.cowlark.fluxengine.arch.macintosh.MacintoshDecoder;
import com.cowlark.fluxengine.arch.micropolis.MicropolisDecoder;
import com.cowlark.fluxengine.arch.mx.MxDecoder;
import com.cowlark.fluxengine.arch.northstar.NorthstarDecoder;
import com.cowlark.fluxengine.arch.rolandd20.RolandD20Decoder;
import com.cowlark.fluxengine.arch.smaky6.Smaky6Decoder;
import com.cowlark.fluxengine.arch.tartu.TartuDecoder;
import com.cowlark.fluxengine.arch.tids990.Tids990Decoder;
import com.cowlark.fluxengine.arch.victor9k.Victor9kDecoder;
import com.cowlark.fluxengine.arch.zilogmcz.ZilogMczDecoder;
import com.cowlark.fluxengine.config.ConfigProto;
import com.cowlark.fluxengine.core.FluxEngineException;
import com.cowlark.fluxengine.decoders.Decoder;
import com.cowlark.fluxengine.decoders.DecoderProto;

/**
 * The Arch class, ported from arch/arch.{h,cc}.
 */
public final class Arch
{
    private Arch()
    {
    }

    public static Decoder createDecoder(ConfigProto config)
    {
        if (!config.hasDecoder())
            throw new FluxEngineException("no decoder configured");
        return createDecoder(config.getDecoder());
    }

    public static Decoder createDecoder(DecoderProto config)
    {
        switch (config.getFormatCase())
        {
            case AGAT:
                return new AgatDecoder(config);
            case AESLANIER:
                return new AesLanierDecoder(config);
            case AMIGA:
                return new AmigaDecoder(config);
            case APPLE2:
                return new Apple2Decoder(config);
            case BROTHER:
                return new BrotherDecoder(config);
            case C64:
                return new Commodore64Decoder(config);
            case F85:
                return new DurangoF85Decoder(config);
            case FB100:
                return new Fb100Decoder(config);
            case IBM:
                return new IbmDecoder(config);
            case MACINTOSH:
                return new MacintoshDecoder(config);
            case MICROPOLIS:
                return new MicropolisDecoder(config);
            case MX:
                return new MxDecoder(config);
            case NORTHSTAR:
                return new NorthstarDecoder(config);
            case ROLANDD20:
                return new RolandD20Decoder(config);
            case SMAKY6:
                return new Smaky6Decoder(config);
            case TARTU:
                return new TartuDecoder(config);
            case TIDS990:
                return new Tids990Decoder(config);
            case VICTOR9K:
                return new Victor9kDecoder(config);
            case ZILOGMCZ:
                return new ZilogMczDecoder(config);
            default:
                throw new FluxEngineException("no decoder specified");
        }
    }
}